### ISSUE TITLE
Bug: Adding divs around elements breaks the grouping list

### ISSUE THREAD

**Issue by joshferrell** [guidepup/virtual-screen-reader#67]
# Issue

Setup:

- Node version: 20.9
- Virtual Screen Reader version: 0.24.0
- Environment: vitest

## Details

It appears that when a div is surrounding an input in a fieldset, the grouping fails to add the items to the set. I verified this worked in VoiceOver as expected.

I'm going to start looking into it this week, but wanted to create the issue first in case you had a suggestion on where to start.

[CodeSandbox reproduction](https://codesandbox.io/p/devbox/infallible-wildflower-y2ckq7?file=%2Fpackage.json%3A7%2C21)

**Comment by cmorten** [comment]
I’m getting:

> Devbox not found
>
> It's likely that the Devbox you're trying to access doesn't exist or you don't have the required permissions to access it.

on the link you’ve shared. Can you check the link/ permissions?

**Comment by joshferrell** [comment]
@cmorten Sorry, I've updated the sandbox permissions now.

**Comment by cmorten** [comment]
> @cmorten Sorry, I've updated the sandbox permissions now.

Amazing thanks!

Gut feel is that we might need to extend the group and position algorithms to ignore `generic` role elements as well as elements with role `presentation` or `none`.

Would be good to check if this is implied/required by any spec, and also what NVDA (and other screen readers) does in this scenario.

**Comment by cmorten** [comment]
See https://github.com/guidepup/virtual-screen-reader/blob/main/src/getNodeAccessibilityData/getAccessibleAttributeLabels/getLabelFromImplicitHtmlElementValue/index.ts#L128 for the entrypoint for calculating the set for the position/set size.

**Comment by joshferrell** [comment]
Still wrapping my head around this problem (could be wrong here), but it seems to me that the the modification needs to be in the accessibility tree flattening. Once it gets to the function you linked the tree is separating out each radio element individually to the fieldset. I would imagine deleting generic roles but including non-generic children in the flattening process would fix it? But I'm new to this codebase so may be missing something.

**Comment by cmorten** [comment]
That would one option, would need to check that omitting generic role elements doesn’t result in missing any information… off top of my head it might be ok

**Comment by jlp-craigmorten** [comment]
Reading around, the issue here is actually specific to radio buttons afaik - they have a special grouping behaviour:

> [radio](https://www.w3.org/TR/core-aam-1.2/#role-map-radio) role, with the [aria-checked](https://www.w3.org/TR/core-aam-1.2/#ariaCheckedTrue) state set to "true" if the element's [checkedness](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-checked) is true, or "false" otherwise. With [aria-setsize](https://www.w3.org/TR/core-aam-1.1/#ariaSetsize) value reflecting number of type=radio input elements within the [radio button group](https://html.spec.whatwg.org/multipage/input.html#radio-button-group) and [aria-posinset](https://www.w3.org/TR/core-aam-1.1/#ariaPosinset) value reflecting the elements position within the [radio button group](https://html.spec.whatwg.org/multipage/input.html#radio-button-group).

REF: https://www.w3.org/TR/html-aam-1.0/#el-input-radio

Where the linked HTML spec section reads:

> The radio button group that contains an [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) element a also contains all the other [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) elements b that fulfill all of the following conditions:
>
> - The [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) element b's [type](https://html.spec.whatwg.org/multipage/input.html#attr-input-type) attribute is in the [Radio Button](https://html.spec.whatwg.org/multipage/input.html#radio-button-state-(type=radio)) state.
> - Either a and b have the same [form owner](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#form-owner), or they both have no [form owner](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#form-owner).
> - Both a and b are in the same [tree](https://dom.spec.whatwg.org/#concept-tree).
> - They both have a [name](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-name) attribute, their [name](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-name) attributes are not empty, and the value of a's [name](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-name) attribute equals the value of b's [name](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-name) attribute.

REF: https://html.spec.whatwg.org/multipage/input.html#radio-button-group

E.g. you can do madness like...

```html
   <main>
        <section>
          <input type="radio" name="example" id="option1" />
          <label htmlFor="option1">Option 1</label>
        </section>
        <p>Some interrupting text</p>
        <footer>
          <input type="radio" name="example" id="option2" />
          <label htmlFor="option2">Option 2</label>
        </footer>
    </main>
```

And AT will still recognise the radio buttons as being in the same group due to their `name` matching.

Removing the `name` from your example, i.e.

```html
   <main>
      <fieldset>
        <legend>Example</legend>
        <div>
          <input type="radio" id="option1" />
          <label htmlFor="option1">Option 1</label>
        </div>
        <div>
          <input type="radio" id="option2" />
          <label htmlFor="option2">Option 2</label>
        </div>
      </fieldset>
    </main>
```

No longer yields the group information (for VO, need to check with NVDA).

Similiarly the following also doesn't group without the `<div>` because the radio buttons don't meet the matching `name` criteria:

```html
   <main>
      <fieldset>
        <legend>Example</legend>
          <input type="radio" id="option1" />
          <label htmlFor="option1">Option 1</label>
          <input type="radio" id="option2" />
          <label htmlFor="option2">Option 2</label>
      </fieldset>
    </main>
```

So the solution here is to support the special case conditions for consider radio inputs as the same group rather than applying the default algorithm.
